metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1275,12 +1271,21 @@ public static List<ApiMessageAndVersion> generateActivationRecords(
         }
         return records;
     }
+
     class CompleteActivationEvent implements ControllerWriteOperation<Void> {
         @Override
         public ControllerResult<Void> generateRecordsAndResult() {
-            List<ApiMessageAndVersion> records = generateActivationRecords(log, logReplayTracker.empty(),
-                zkMigrationEnabled, bootstrapMetadata, featureControl);
-            return ControllerResult.atomicOf(records, null);
+            try {
+                List<ApiMessageAndVersion> records = generateActivationRecords(log,
+                    logReplayTracker.empty(),
+                    zkMigrationEnabled,
+                    bootstrapMetadata,
+                    featureControl);
+                return ControllerResult.atomicOf(records, null);
+            } catch (Throwable t) {
+                throw fatalFaultHandler.handleFault("exception while completing controller " +
```

Comment:
I think in most other places, we are not inside an event when we call the fault handler in QuorumController. I guess that doesn't matter since it calls `Exit.halt` synchronously?